src/core/mesh/qgsmesheditor.cpp

```diff
@@ -1482,3 +1491,152 @@ void QgsMeshLayerUndoCommandAdvancedEditing::redo()
+  QgsMeshEditor *meshEditor,
+  const QgsMeshVertex vertex,
+  double tolerance )
+  : QgsMeshLayerUndoCommandMeshEdit( meshEditor )
+  , mVertex( vertex )
+  , mTolerance( tolerance )
+{
+  setText( QObject::tr( "Add vertex inside face with Delaunay refinement" ) );
+}
+
+void QgsMeshLayerUndoCommandAddVertexInFaceWithDelaunayRefinement::redo()
+{
+  if ( !mVertex.isEmpty() )
+  {
+    QgsMeshEditor::Edit edit;
+
+    mMeshEditor->applyAddVertex( edit, mVertex, mTolerance );
+    mEdits.append( edit );
+
+    QList<std::pair<int, int>> sharedEdges = innerEdges( triangularFaces( facesContainingVertex( mMeshEditor->topologicalMesh().mesh()->vertexCount() - 1 ) ) );
+
+    for ( std::pair<int, int> edge : sharedEdges )
+    {
+      if ( mMeshEditor->edgeCanBeFlipped( edge.first, edge.second ) && !mMeshEditor->topologicalMesh().delaunayConditionForEdge( edge.first, edge.second ) )
+      {
+        mMeshEditor->applyFlipEdge( edit, edge.first, edge.second );
+        mEdits.append( edit );
+      }
+    }
+
+    mVertex = QgsMeshVertex();
+  }
+  else
+  {
+    for ( QgsMeshEditor::Edit &edit : mEdits )
+      mMeshEditor->applyEdit( edit );
+  }
+}
+
+QSet<int> QgsMeshLayerUndoCommandAddVertexInFaceWithDelaunayRefinement::facesContainingVertex( const int vertexId )
+{
+  QList<int> facesAroundVertex;
+
+  for ( int i = 0; i < mMeshEditor->topologicalMesh().mesh()->faceCount(); i++ )
+  {
+    QgsMeshFace face = mMeshEditor->topologicalMesh().mesh()->face( i );
+
+    if ( face.contains( vertexId ) )
+    {
+      facesAroundVertex.push_back( i );
+    }
+  }
+
+  QSet<int> vertexIndexes;
+
+  for ( int faceIndex : facesAroundVertex )
+  {
+    QgsMeshFace face = mMeshEditor->topologicalMesh().mesh()->face( faceIndex );
+
+    for ( int i = 0; i < face.count(); i++ )
+    {
+      vertexIndexes.insert( face.at( i ) );
+    }
+  }
+
+  // faces that have at least one common vertex with newly added faces
+  QSet<int> selectedFaces;
+
+  for ( int i = 0; i < mMeshEditor->topologicalMesh().mesh()->faceCount(); i++ )
+  {
+    const QgsMeshFace face = mMeshEditor->topologicalMesh().mesh()->face( i );
+
+    for ( int j = 0; j < face.count(); j++ )
+    {
+      for ( int vertex : vertexIndexes )
+      {
+        if ( face.contains( vertex ) )
+        {
+          selectedFaces.insert( i );
+          break;
+        }
+      }
+    }
+  }
+  return selectedFaces;
+}
+
+
+QSet<int> QgsMeshLayerUndoCommandAddVertexInFaceWithDelaunayRefinement::triangularFaces( const QSet<int> &faces )
+{
+  QSet<int> triangularFaces;
+
+  for ( int faceIndex : faces )
+  {
+    if ( mMeshEditor->topologicalMesh().mesh()->face( faceIndex ).count() == 3 )
+    {
+      triangularFaces.insert( faceIndex );
+    }
+  }
+
+  return triangularFaces;
+}
+
+QList<std::pair<int, int>> QgsMeshLayerUndoCommandAddVertexInFaceWithDelaunayRefinement::innerEdges( const QSet<int> &faces )
```

Comment:
Do we actually need to check which are the inner edges?
Wouldn't outer edges simply fail the `mMeshEditor->edgeCanBeFlipped()` check?

Reply:
Yeh, we need to check for inner edges, as those are inner edges of our area of interest (neighbors to the second level of the face). The outer edges might pass `edgeCanBeFlipped()`, but according to the description, we should not consider those.